### PR TITLE
Prevent analytics collector from missing events when tower is under stress

### PR DESCRIPTION
**Pull request by jladdjr**
Adapts https://github.com/ansible/awx/pull/9358#issuecomment-797747525 to work in the context of awx using partitioned job events tables (see https://github.com/ansible/awx/issues/9039 and https://github.com/ansible/awx/pull/9341).

Relevant discussions:
https://github.com/ansible/tower/issues/4478
and specifically the comment here: https://github.com/ansible/tower/issues/4478#issuecomment-771471086

### CODE REVIEW

**Comment by jladdjr** [comment]
@jbradberry @ryanpetrello @Ladas @kdelee 

Going to try to start tinkering on this

**Comment by jladdjr** [comment]
@jbradberry reminded me of one thorny issue here - what to do about the fact that we can't count on id's to be unique across all partitions.

* Instead of `id` being a primary key on jobevent tables
* .. we now how a _composite key_ consisting of id and job_created (the field we're using as a basis for partitioning)
* So we know that an id will be unique _within_ a given partition
* .. but in a past meeting @Ladas shared that he's seen a postgres sequence get tripped up and use the same sequence value across multiple partitions

**Comment by Ladas** [comment]
@jladdjr I think we talked that we'd use `modified` instead of id for 4.0 ?

**Comment by jladdjr** [comment]
> I think we talked that we'd use modified instead of id for 4.0 ?

@Ladas I think I forgot what our reasoning was for using `modified` when I context switched to looking at things in perfscale. Could you outline how you think we should do this again?

**Comment by jladdjr** [comment]
I found some notes on this (😌 ) - based on what I wrote:
* we want to try to ship out job events for entire partitions
* it's okay for analytics to get duplicate data
* we will ship data if a job associated with a given partition has been modified after we last shipped events

@Ladas @ryanpetrello does that sound like the approach we're wanting to try?

**Comment by jladdjr** [comment]
rebased the latest version of @jbradberry 's [fix](https://github.com/ansible/awx/pull/9358/commits)
onto the HEAD of the current [db_partition branch](https://github.com/ansible/awx/pull/9341)

**Comment by jladdjr** [comment]
cherry-picking https://github.com/ansible/awx/pull/9825/files until this lands in devel

**Comment by jladdjr** [comment]
@chrismeyersfsu @Ladas @kdelee - this PR now addresses two areas that @Ladas and I chatted about a little while ago:

**1. search for job events based on their `modified` time, instead of `created` time**

job events can be created, but take a little while to actually land in the db. The `modified` should much more accurately reflect the time the event was committed to the db, so this change should help us catch events that lag a bit on the way to the db.

**2. when collecting job events up to the horizon, copy events by _partition_ when possible**

When we know that we need to scoop up 4 weeks worth of events, and all events reside in partitions, we can go ahead and get the events one partition at a time.

**Comment by jladdjr** [comment]
Taking a step back, I wanted to mention that before doing any of the work here, I ran a test on my local dev env to see how quickly `awx-manage gather_analytics` ran with 2 million events. I compared `devel` to the `db_partition` branch. I wanted to see if we got a performance bump just by using partitioned tables before adding the tweaks in this PR. I saw a significant boost in performance:

```
# devel

bash-4.4# time awx-manage gather_analytics --since 2021-03-01 --ship
2021-04-05 23:10:43,835 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_0.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_0.tar.gz
2021-04-05 23:10:44,008 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_1.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_1.tar.gz
2021-04-05 23:10:44,010 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_2.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_2.tar.gz
2021-04-05 23:10:44,011 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_3.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-230404+0000_3.tar.gz

real    6m44.859s
user    0m28.172s
sys     0m3.412s



# db_partition

docker exec -it -u 0 tools_awx_1 bash
bash-4.4# time awx-manage gather_analytics --since 2021-03-01 --ship
2021-04-05 23:45:18,150 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_0.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_0.tar.gz
2021-04-05 23:45:18,199 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_1.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_1.tar.gz
2021-04-05 23:45:18,200 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_2.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_2.tar.gz
2021-04-05 23:45:18,201 INFO     [-] awx.main.analytics /tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_3.tar.gz
/tmp/00000000-0000-0000-0000-000000000000_2021-04-05-234132+0000_3.tar.gz

real    3m50.094s
user    0m25.274s
sys     0m3.195s
```

`(3*60+50)/(6*60+45) = 56.8%`

TLDR: It took 43% less time to gather analytics for 2m events using the partitioned branch.

**Comment by jladdjr** [comment]
After writing this PR, I was curious to see how much time we were saving by collecting job events one partition at a time when we have to collect all events up to the horizon. In this case, I saw a very modest improvement:

**db_partition_analytics, getting all events up to horizon _selecting by id_ (existing behavior)**

```
real    3m57.436s
user    0m23.276s
sys     0m2.265s
```

**db_partition_analytics, getting all events up to horizon _using partitions_**

```
real    3m38.451s
user    0m23.431s
sys     0m3.259s
```

`(3*60+38)/(3*60+57) = 92%`

I only saw an 8% improvement in this case.

---

I wondered why we didn't see a bigger performance boost, so I checked some queries:

Asking pg to get rows by id:
```
awx=# explain select * from main_jobevent where id > 724022 and id < 1071637;
                                                       QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------
 Append  (cost=0.42..87267.60 rows=347622 width=1776)
   ->  Index Scan using main_jobevent_20210401_18_pkey on main_jobevent_20210401_18  (cost=0.42..8.44 rows=1 width=1774)
         Index Cond: ((id > 724022) AND (id < 1071637))
   ->  Index Scan using main_jobevent_20210401_19_pkey on main_jobevent_20210401_19  (cost=0.42..8.44 rows=1 width=1773)
         Index Cond: ((id > 724022) AND (id < 1071637))
   ->  Index Scan using main_jobevent_20210401_20_pkey on main_jobevent_20210401_20  (cost=0.42..8.44 rows=1 width=1774)
         Index Cond: ((id > 724022) AND (id < 1071637))
   ->  Seq Scan on main_jobevent_20210401_21  (cost=0.00..85462.24 rows=347614 width=1776)
         Filter: ((id > 724022) AND (id < 1071637))
   ->  Index Scan using main_jobevent_20210401_22_pkey on main_jobevent_20210401_22  (cost=0.42..8.44 rows=1 width=1775)
         Index Cond: ((id > 724022) AND (id < 1071637))
   ->  Index Scan using main_jobevent_20210401_23_pkey on main_jobevent_20210401_23  (cost=0.42..8.44 rows=1 width=1774)
         Index Cond: ((id > 724022) AND (id < 1071637))
   ->  Index Scan using main_jobevent_20210402_00_pkey on main_jobevent_20210402_00  (cost=0.42..8.44 rows=1 width=1771)
         Index Cond: ((id > 724022) AND (id < 1071637))
   ->  Index Scan using main_jobevent_20210402_01_pkey on main_jobevent_20210402_01  (cost=0.28..8.30 rows=1 width=1773)
         Index Cond: ((id > 724022) AND (id < 1071637))
   ->  Index Scan using main_jobevent_20210402_16_pkey on main_jobevent_20210402_16  (cost=0.29..8.30 rows=1 width=1773)
         Index Cond: ((id > 724022) AND (id < 1071637))
(19 rows)
```

Asking pg to get all rows in a partition:
```
awx=# explain select * from main_jobevent_20210401_21;;
                                     QUERY PLAN
-------------------------------------------------------------------------------------
 Seq Scan on main_jobevent_20210401_21  (cost=0.00..83724.16 rows=347616 width=1776)
(1 row)
```

It looks like selecting by id is only a _little_ slower because postgres does an index scan on all of the partitions, but then quickly says "nope, nothing I need here" and moves on. The cost of those index scans seems pretty small. In other words, even though we're asking postgres to search through all partitions, it seems smart enough that it hardly bothers with the partitions that don't have matching ids.

So. This makes me wonder how much we want this optimization. In order to slice by partition, I abuse @jbradberry 's existing pattern a bit -- using a generator that returns partition names instead of id ranges.

Thoughts?

File: awx/main/analytics/collectors.py [hunk 1]
```diff
@@ -76,7 +78,51 @@ def events_slicing(key, since, until):
     lower = since or last_gather or horizon
     if not since and last_entries.get(key):
         lower = horizon
-    pk_values = models.JobEvent.objects.filter(created__gte=lower, created__lte=until).aggregate(Min('pk'), Max('pk'))
+    partition_epoch = get_event_partition_epoch()
+
+    def get_partitions():
```
Comment (jladdjr) [inline awx/main/analytics/collectors.py:83]: These helper functions would likely end up in a common library once the job event clean-up changes land

File: awx/main/analytics/collectors.py [hunk 2]
```diff
@@ -76,7 +78,51 @@ def events_slicing(key, since, until):
...
+        dt = datetime.datetime.strptime(dt_str, '%Y%m%d_%H').replace(tzinfo=pytz.UTC)
+        return dt > horizon
+
+    # TODO: Make default once unpartitioned tables are no longer supported
+    # if we are collecting job events starting at the horizon
+    # and all job events are stored in partitions,
+    # slice by partition
+    if lower == horizon and lower > partition_epoch:
+        for partition in get_partitions():
+            if partition_within_horizon(partition, horizon):
```
Comment (jladdjr) [inline awx/main/analytics/collectors.py:111]: I'm abusing @jbradberry 's existing generator pattern a bit here, returning partition names instead of ids.

Comment (jladdjr) [reply awx/main/analytics/collectors.py:111]: Despite this abuse, nothing in `analytics/core.py` needs to change. That code can be pretty agnostic about what's happening here.. I _think_.

File: awx/main/analytics/collectors.py [hunk 3]
```diff
@@ -76,7 +78,51 @@ def events_slicing(key, since, until):
...
+                yield (partition, None)
+        return
+
+    # JobEvent.modified index was created at the partition epoch
+    # Events created before this are in a separate table that does not
+    # have this index.
+    if lower >= partition_epoch:
```
Comment (jladdjr) [inline awx/main/analytics/collectors.py:118]: I tried to think this through carefully. Basically, if you're an old job event, we look you up using the `created` field. If your new, we use the `modified` field. If our search spans both old and new fields, we adopt both approaches. 

There's probably a more elegant way to write this code-wise, but I didn't want to _always_ potentially do two queries for the sake of reducing the code here.

File: awx/main/analytics/collectors.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
 import os.path
 import platform
 import distro
+import pytz
```
Comment (jbradberry) [inline awx/main/analytics/collectors.py:7]: `utc` is already included in `django.utils.timezone`, it would probably be better to just add it to that import below.

File: awx/main/analytics/collectors.py [hunk 5]
```diff
@@ -76,7 +78,51 @@ def events_slicing(key, since, until):
...
+    def partition_within_horizon(partition, horizon):
+        dt_str = partition[14:]
+        dt = datetime.datetime.strptime(dt_str, '%Y%m%d_%H').replace(tzinfo=pytz.UTC)
+        return dt > horizon
+
+    # TODO: Make default once unpartitioned tables are no longer supported
+    # if we are collecting job events starting at the horizon
+    # and all job events are stored in partitions,
+    # slice by partition
+    if lower == horizon and lower > partition_epoch:
```
Comment (jbradberry) [inline awx/main/analytics/collectors.py:109]: If a user passes in a `since` parameter or if we are making use of the recorded last successful ship, surely that is just as worthy of partition optimization?

File: awx/main/analytics/collectors.py (hunk 3 above)
Comment (jbradberry) [reply awx/main/analytics/collectors.py:118]: Presumably even your legacy data now has an index on `modified`?  Use it in all cases, there's no other reason to use `created`.

Comment (ryanpetrello) [reply awx/main/analytics/collectors.py:118]: Nope, the old table doesn't have a `modified` index.

**Comment by jbradberry** [comment]
@jladdjr All right, I think there are significant problems here:

1. Slicer functions need to return consistent, comparable data types for the interval endpoints.  It looks to me like you now can return either `None` (comparable with nothing, not even itself), `int`, or `str`, which cannot be compared with each other.  The first entry in each yielded tuple, if I remember correctly, can be a meaningless value if need be as long as the collector function doesn't naively pass it on in the SQL, but the second value is what ultimately gets recorded on a successful ship.  It absolutely needs to be consistent and comparable with inequality operators.
2. You are conflating what the slicer needs to do with what the collector needs to do.  I think if your slicer just returned datetime intervals intended to be used to query against `modified`, there is still plenty of room to do your collect-whole-partitions optimization within the collector.  Or, perhaps we could be using tuples of partition + datetime as the endpoints in order to make sure that the intervals constructed wind up not always breaking up partitions.  That might work, I think, as long as the tuples involved always can be compared using the inequality operators, and if the semantics of that comparison makes sense.  I wonder what is supposed to happen, though, if a partition that we already shipped gets a few more events added after it.  I can't really think that through.

Honestly though, I wonder if the complexity of trying to be partition-aware is even worth it.  It seems like just using `modified` is an incredibly easy change (you could probably drop `events_slicing` entirely!) and will transparently do most of what we would want.

**Comment by Ladas** [comment]
@jladdjr still going through this, first thing I noticed is that we'll probably need a bigger dataset with more diverse data to test this.

E.g. for bigger clusters, I assume that there will be hundreds of partitions + we can expect the interval will be spread over many partitions -> sequential scan of several partitions for 1 slice. So the partition scan might be causing much bigger speedup, doing this for the initial 4 weeks back, since that will be a lot of slices.

In any way, even if we decide we don't need optimizations now, we should probably track how long the collection takes? So we know if we need to optimize (e.g. seeing some collection took a day)? Sending the time of the collection as part of  https://github.com/ansible/tower/issues/4793 ?

**Comment by jladdjr** [comment]
rebased

File: awx/main/analytics/collectors.py (hunk 5 above)
Comment (jladdjr) [reply awx/main/analytics/collectors.py:111]: @ladas - this seems like a good point ^. Do you think we should just adopt this approach in general?

**Comment by jladdjr** [comment]
> Honestly though, I wonder if the complexity of trying to be partition-aware is even worth it. It seems like just using modified is an incredibly easy change (you could probably drop events_slicing entirely!) and will transparently do most of what we would want.

@jbradberry - this is pretty much what I was thinking at the end of yesterday. I wanted to show what the partition-based slicing _would_ look like, though, so we could talk about it's merits before deciding against it, though.

**Comment by jladdjr** [comment]
> @jladdjr still going through this, first thing I noticed is that we'll probably need a bigger dataset with more diverse data to test this.

@Ladas - I agree. I wanted to do a quick check locally before running things at scale. (Didn't want to run things at scale if we found obvious issues w/ a smaller set). Whatever options we think are worth investigating, I think we should def. exercise them at scale as our next step.

**Comment by jladdjr** [comment]
> In any way, even if we decide we don't need optimizations now, we should probably track how long the collection takes?

@Ladas Yes, definitely. I'm not sure if we'll manage to fit that in here, but yes we should do this.

**Comment by jladdjr** [comment]
we moved the db_partition branch from my repo over to ansible's. closing this PR and reopening a [new one](https://github.com/ansible/awx/pull/9856) that is based on ansible/db_partition
